packages/backend-common/src/appContext/ApplicationContext.ts

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Logger } from 'winston';
+import { Container, interfaces } from 'inversify';
+import express from 'express';
+import { AnyDependencyConfig, Dependency } from '@backstage/app-context-common';
+
+export interface ApplicationContext {
+  getChildContext(identifier: string): ApplicationContext;
+
+  getContainer(): interfaces.Container;
+
+  get<T>(dep: Dependency<T>): T;
+
+  createBoundPlugin(options: BoundPluginOptions): BoundPlugin;
+
+  forPlugin(plugin: string): ApplicationContext;
```

Comment:
Gotta say I still highly doubt that we actually need this and that it's worth it. We've peeked at this dual-layer approach before and it made things a lot more complex very quickly. My hunch is that it's much easier to figure out workaround for cases where we want to have plugin-specific instances instead. Getting into some more concrete discussion further down tho

Reply:
Thinking from an architectural point of view, the way I think about this is that this should be something that we add in. Let me know if I'm missing other options/use cases here.

We have few different options in here but the downsides of others are something we'd want to try to avoid.

1. Creating a single container, without the `forPlugin` call at all
   * All plugins/modules would be using the same container and would have access to (all) dependencies within the monolith.
   * We'd loose module isolation and would leak module specific deps so they are usable elsewhere. BadgeFactories as bound within this PR would be usable in tech-insights-backend module without adding the explicit dependency to that badges package. This would give developers the ability to introduce implicit, hard dependencies between modules preventing them from being run individually.
   * Overriding individual dependencies for a single module would be doable (by using named/tagged/whatever bindings from Inversify), if the module allows it. We'd want to allow it in most cases, which would mean all dependency definitions would end up being named/tagged/whatever ability Inversify provides.
2. Creating an individual, separate container for each plugin/module
   * Each plugin would be completely isolated and would handle their own deps. We would construct each container with similar initial structure for the common deps (DBmgr, cachemgr etc.). 
   * Overriding individual deps for plugin/module would need to be done for each each of them individually. 
   * The actual architectural change of this would be a wrapper around the same logic that has previously existed and wouldn't actually bring any benefits from a container, apart from intermodule testing etc. dep-swapping possibilities. In other words, 'What's the point', modules can already use Inversify independently if they want. 
3. This `forPlugin` (naming pending/working title as usual) approach 
   * The "application" (whether it's the modular monolith of the whole Backstage or individual package run individually) would have a root container binding common deps, each module would have their own container, as a child container of the root container, for their own deps. 
   * This would allow modules traverse up the container from child to root if they need common, higher-level common deps (like DBmgr, cachemgr, etc.) and have their own implementations for module level deps.
   * Dependencies within the module would be isolated to the child container, preventing cross module leakage of dependencies
   * Each module would have the ability to have their own isolated module-level override for any of the root dependencies, since Inversify would use the dep from the child before traversing up. This would give modules the possibility to either use a commonly bound dependency or swap the implementation on the module level. Say, you want to override GitHubCredentialsProvider for a single module, you would provide a different DependencyConfig for that one module, all others could use a previously registered root module one.

Reply:
Yep, I think (2) or (3) are the only real options, unless we want to completely change how we provide access to DB resources and the like.

> Each plugin would be completely isolated and would handle their own deps.

Exactly, I think that's really good :grin:

FE and BE are very different in that for BE any cross-plugin communication really needs to happen over the network. We may have clients being shared and injected, but that's hopefully about it. Because of this isolation I feel that the top-level context only ends up serving as a utility that holds a couple of factories, which might be simpler, clearer, and more flexible to express with regular boring code, similar to our current env setup function.

One addition that could simplify things a lot is to inject a `PluginMeta` dependency, or w/e it would be called. The main (maybe only) role of it would be to allow access to the plugin ID within the plugin container. That way you end up not really needing the `forPlugin` call, you just need to register a factory that declares a dependency on `PluginMeta` and it can use that to spin up the DB or w/e it's trying to do. The factory itself would hole a reference to global state if that's desired for connection pooling, etc.

I think the fact that (3) lets you share dependencies from the root container actually has very little value, you might as well just register a factory in each plugin container instead. With a nice pattern for that I think it should be the same amount of code too, if not less.

> Dependencies within the module would be isolated to the child container, preventing cross module leakage of dependencies

This is of course true for (2) as well, but even more strictly enforced. I think the same goes for the last point of (3) as well.

Reply:
I decoupled the containers in this case by just spinning up a separate container within the ApplicationContext class. It's not ideal so I'll refactor the initialization to a few plain functions and leave that logic out of the AppContext itself, since it's not needed there. Was that something you were thinking of?

Reply:
Resolving this conversation for now. Modified the implementation to have a top level manager that spins up individual, isolated app contexts that can be passed into plugins.